### PR TITLE
Add replicas field for HA deployments

### PR DESCRIPTION
**Pull request by alaxkar**
## Summary

Adds optional `replicas` field to `GenericDeploymentSpec` for configuring replica counts per KEDA component (operator, metricsServer, admissionWebhooks), enabling HA deployments.

## Changes

- **API**: Added `Replicas *int32` field to `GenericDeploymentSpec`
- **Validation**:
  - Operator: 0-2 replicas (allows 0 for maintenance scenarios, max 2 for HA)
  - MetricsServer: minimum 1 replica, no maximum limit
  - AdmissionWebhooks: minimum 1 replica, no maximum limit
- **Controller**: Component-specific validation enforced during reconciliation
- **Tests**: Unit tests for validation boundary cases
- **CRD**: Enforces `minimum: 0` at API server level

## Example Usage

```yaml
apiVersion: keda.sh/v1alpha1
kind: KedaController
metadata:
  name: keda
spec:
  operator:
    replicas: 2
  metricsServer:
    replicas: 3
  admissionWebhooks:
    replicas: 2
```

## Benefits

- Enables HA configurations for better availability
- Allows scaling operator to 0 for maintenance without affecting other components
- Provides flexibility for different deployment scenarios

## Summary by CodeRabbit

- **New Features**
  - Added configurable replica counts for deployment components, including the operator, metrics server, admission webhooks, and HTTP Add-on deployments.
  - Replica settings now use consistent validation; operator, metrics server, and admission webhook counts must be at least one when specified.

- **Bug Fixes**
  - Configuration and secret updates now refresh controller status without deleting or restarting the metrics server pod.
  - Standardized replica descriptions and validation across deployment configuration schemas.

### CODE REVIEW

File: api/keda/v1alpha1/kedacontroller_types.go [hunk 1]
```diff
@@ -374,6 +380,18 @@ type GenericDeploymentSpec struct {
 	Env []corev1.EnvVar `json:"env,omitempty"`
 }
 
+// Validate validates the GenericDeploymentSpec fields.
+// - Allows replicas >= 0 (0 may be used for maintenance scenarios)
+// - Blocks negative values (invalid)
+// - Allows nil (uses base manifest default)
+// - Component-specific minimums enforced in controller (e.g., metricsServer requires >= 1)
+func (g *GenericDeploymentSpec) Validate() error {
```
Comment (linkvt) [inline api/keda/v1alpha1/kedacontroller_types.go:388]: Is this validation really needed considering the `Minimum=0` above? I don't think values < 0 are possible via Kubernetes CRs.

File: internal/controller/keda/kedacontroller_controller.go [hunk 2]
```diff
@@ -412,6 +414,66 @@ func (r *KedaControllerReconciler) Reconcile(ctx context.Context, req ctrl.Reque
...
+	if err := instance.Spec.Operator.Validate(); err != nil {
+		logger.Error(err, "Invalid Operator configuration")
+		status.MarkInstallFailed(fmt.Sprintf("Invalid Operator configuration: %v", err))
+		if updateErr := util.UpdateKedaControllerStatus(ctx, r.Client, instance, status); updateErr != nil {
+			err = fmt.Errorf("got error: %s and then another: %s", err, updateErr)
+		}
+		return ctrl.Result{}, err
+	}
+	// Operator-specific validation: max 2 replicas
+	if instance.Spec.Operator.Replicas != nil && *instance.Spec.Operator.Replicas > 2 {
```
Comment (linkvt) [inline internal/controller/keda/kedacontroller_controller.go:427]: Considering these validations, should we apply them via CEL expressions to avoid all of this code? I was thinking, better prevent invalid input from reaching the reconciliation loop instead of having to handle it.

File: config/default/kustomization.yaml [hunk 3]
```diff
@@ -45,7 +45,9 @@ resources:
...
 - ../manager
 labels:
-- pairs:
+- includeSelectors: true
+  pairs:
     app.kubernetes.io/part-of: keda-olm-operator
     app.kubernetes.io/version: main
-  includeSelectors: true
+commonLabels:
+  app.kubernetes.io/version: main
```
Comment (linkvt) [inline config/default/kustomization.yaml:53]: Were these changes and in config/manager/kustomization.yaml included on purpose?

**Review by linkvt — COMMENTED** [review]
Thanks for the PR, I left a few comments and will trigger a copilot review as well.
Maybe a general question, is there a requirement to be this strict, i.e. to fail on 3+ operator pods? I understand that more than 2 doesn't really add much value but just wondering how strict we want to be.

**Review by Fedosin — COMMENTED** [review]
Thanks for putting this together. HA support for the KEDA components is definitely worth having, and putting `replicas` on the shared `GenericDeploymentSpec` and reusing `transform.ReplaceReplicas` was the right call. I pulled the branch down and it builds cleanly.

There are a couple of things I think we should sort out before this merges, though.

**Multi-replica metrics server breaks cert rotation.** `util.DeleteMetricsServerPod` refuses to do anything unless there's exactly one pod running, and that's the function we call when the certs Secret or the CA bundle ConfigMap changes. So as soon as someone sets `metricsServer.replicas: 2`, cert rotation silently stops working. I think this needs to become a rollout restart rather than deleting a single pod.

**`replicas` collides with the HTTP Add-on components.** `HTTPAddonOperatorSpec`, `HTTPAddonInterceptorSpec` and `HTTPAddonScalerSpec` all define their own `Replicas` field on top of the embedded `GenericDeploymentSpec`. Go and controller-gen don't agree on which one takes precedence, so the CRD schema drifts from the code. You can see it in the diff, where the descriptions on those three fields got replaced with the generic one. I'd just drop the three duplicate fields and let them inherit from the shared spec.

On validation, I mostly agree with what @linkvt already said:

* `Validate()` can never fail while `Minimum=0`, so it's effectively dead code, and all eight new tests only exercise that dead method. None of them cover the rules this PR actually adds.
* With `minimum: 0` in the schema and the "must be >= 1" check living in the reconcile loop, the API server will happily accept `metricsServer.replicas: 0`, and the user only finds out something's wrong from a status condition later. CEL validation is the right fix here.
* On the strictness question @linkvt raised: I'd drop the operator cap entirely. The operator runs with leader election, so extra replicas are just idle standbys. Rejecting 3 blocks anyone who wants a standby per AZ, and the limit won't age well if KEDA ever starts sharding work across replicas.
* I'd also settle on one consistent story for zero replicas. Right now `operator.replicas: 0` is a feature while `metricsServer.replicas: 0` is an error, which feels arbitrary. And scaling the operator to zero isn't side-effect free the way the description suggests, since that deployment handles cert rotation and serves the gRPC endpoint the metrics adapter reads from.

Smaller things:

* The six validation blocks in `Reconcile` are a lot of copy-paste. A single helper would cover all of them, and you'd be able to drop the `//nolint:gocyclo` on `Reconcile` as well. Worth seeing whether the one on `installMetricsServer` can be avoided too, rather than silencing the linter in two places.
* `go.mod` picked up ginkgo v1, but the rest of the repo is on v2.
* The two kustomization.yaml changes look like stray `make` output, agreeing with @linkvt there.
* The README documents all the other `GenericDeploymentSpec` fields per component, so `replicas` should be added there too.
* There's no e2e coverage for any of this. `test/e2e/upgrade_test.go` already sets `Replicas`, so there's a natural place to add it.

File: api/keda/v1alpha1/kedacontroller_types.go (hunk 1 above)
Comment (alaxkar) [reply api/keda/v1alpha1/kedacontroller_types.go:388]: You are right, negative value would be rejected in the CR. I made it intentionally, I was unaware about the behaviour that negative value would be rejected by default. I will remove the validation and check whether negative values are rejected in the CR. Thank you!!

File: config/default/kustomization.yaml (hunk 3 above)
Comment (alaxkar) [reply config/default/kustomization.yaml:53]: Actually no these are artifacts from running make deploy & make bundle in both of the files while testing. It is not intentional. Should I revert these?

Comment (linkvt) [reply config/default/kustomization.yaml:53]: Yes, I think we should revert these. I didn't look into it yet but IMO these files shouldn't get dirty at all during normal development, but that's a future issue maybe 😀

File: internal/controller/keda/kedacontroller_controller.go (hunk 2 above)
Comment (alaxkar) [reply internal/controller/keda/kedacontroller_controller.go]: Thank you, Done! I have replaced controller validation with CEL rules. Validation now happens at the API server level for immediate feedback to the user.

File: config/default/kustomization.yaml (hunk 3 above)
Comment (alaxkar) [reply config/default/kustomization.yaml]: Reverted both files now. Thanks

**Comment by alaxkar** [comment]
Thanks you @Fedosin for your review.
> There are a couple of things I think we should sort out before this merges, though.
> 
> **Multi-replica metrics server breaks cert rotation.** `util.DeleteMetricsServerPod` refuses to do anything unless there's exactly one pod running, and that's the function we call when the certs Secret or the CA bundle ConfigMap changes. So as soon as someone sets `metricsServer.replicas: 2`, cert rotation silently stops working. I think this needs to become a rollout restart rather than deleting a single pod.

Done, Fixed with rolling restart approach.

> **`replicas` collides with the HTTP Add-on components.** `HTTPAddonOperatorSpec`, `HTTPAddonInterceptorSpec` and `HTTPAddonScalerSpec` all define their own `Replicas` field on top of the embedded `GenericDeploymentSpec`. Go and controller-gen don't agree on which one takes precedence, so the CRD schema drifts from the code. You can see it in the diff, where the descriptions on those three fields got replaced with the generic one. I'd just drop the three duplicate fields and let them inherit from the shared spec.

For HTTP Add-on collision, I have removed duplicate fields and corrected the code with

> On validation, I mostly agree with what @linkvt already said:
> 
> * `Validate()` can never fail while `Minimum=0`, so it's effectively dead code, and all eight new tests only exercise that dead method. None of them cover the rules this PR actually adds.
> * With `minimum: 0` in the schema and the "must be >= 1" check living in the reconcile loop, the API server will happily accept `metricsServer.replicas: 0`, and the user only finds out something's wrong from a status condition later. CEL validation is the right fix here.
> * On the strictness question @linkvt raised: I'd drop the operator cap entirely. The operator runs with leader election, so extra replicas are just idle standbys. Rejecting 3 blocks anyone who wants a standby per AZ, and the limit won't age well if KEDA ever starts sharding work across replicas.
> * I'd also settle on one consistent story for zero replicas. Right now `operator.replicas: 0` is a feature while `metricsServer.replicas: 0` is an error, which feels arbitrary. And scaling the operator to zero isn't side-effect free the way the description suggests, since that deployment handles cert rotation and serves the gRPC endpoint the metrics adapter reads from.
> 
> Smaller things:
> 
> * The six validation blocks in `Reconcile` are a lot of copy-paste. A single helper would cover all of them, and you'd be able to drop the `//nolint:gocyclo` on `Reconcile` as well. Worth seeing whether the one on `installMetricsServer` can be avoided too, rather than silencing the linter in two places.
> * `go.mod` picked up ginkgo v1, but the rest of the repo is on v2.
> * The two kustomization.yaml changes look like stray `make` output, agreeing with @linkvt there.
> * The README documents all the other `GenericDeploymentSpec` fields per component, so `replicas` should be added there too.
> * There's no e2e coverage for any of this. `test/e2e/upgrade_test.go` already sets `Replicas`, so there's a natural place to add it.


That's really a good point considering the high availability across AZ's. Now, I will remove the max=2 operator limit to allow multi-AZ, and have responded to the Vincent's suggestions with the changes. :)

**Comment by alaxkar** [comment]
Thanks for the thorough reviews! I've addressed all the feedback:

**Changes:**
- Implemented CEL validation as suggested by @linkvt
- Fixed multi-replica cert rotation issue (@Fedosin, @Copilot) 
- Removed duplicate Replicas fields from HTTP Add-on specs (@Copilot)
- Cleaned up validation code and reverted kustomization artifacts
- Removed operator max=2 replica limit to allow multi-AZ deployments (@Fedosin, @linkvt)

**Question:**
Should operator require replicas >= 1 for consistency with metricsServer/admissionWebhooks, or keep allowing 0?

Let me know your preference.

File: api/keda/v1alpha1/kedacontroller_types.go [hunk 4]
```diff
@@ -70,6 +70,7 @@ type KedaServiceAccountSpec struct {
 	Labels map[string]string `json:"labels,omitempty"`
 }
 
+// +kubebuilder:validation:XValidation:rule="!has(self.replicas) || self.replicas >= 0",message="operator replicas must be >= 0"
```
Comment (linkvt) [inline api/keda/v1alpha1/kedacontroller_types.go]: nit: This rule is not needed as the replicas already has a validation:
```go
	// +kubebuilder:validation:Minimum=0
```

File: api/keda/v1alpha1/kedacontroller_types.go [hunk 5]
```diff
@@ -314,6 +316,10 @@ type HTTPAddonStatus struct {
 }
 
 type GenericDeploymentSpec struct {
+	// Number of replicas for the deployment
+	// +optional
+	// +kubebuilder:validation:Minimum=0
+	Replicas *int32 `json:"replicas,omitempty"`
```
[the remark this answers is not in the document]
Comment (linkvt) [reply api/keda/v1alpha1/kedacontroller_types.go:310]: Resolved

File: go.mod [hunk 6]
```diff
@@ -13,6 +13,7 @@ require (
 	github.com/go-logr/logr v1.4.3
 	github.com/manifestival/controller-runtime-client v0.4.0
 	github.com/manifestival/manifestival v0.7.3-0.20230801201407-f20c69532c27
+	github.com/onsi/ginkgo v1.16.5
```
[the remark this answers is not in the document]
Comment (linkvt) [reply go.mod]: @alaxkar I think this is still unresolved.

File: internal/controller/keda/util/util.go [hunk 7]
```diff
@@ -39,29 +44,46 @@ func CalculateSecretedDataCheckSum(m map[string][]byte) string {
 	return fmt.Sprintf("%x", md5.Sum([]byte(data)))
 }
 
+// DeleteMetricsServerPod triggers a rolling restart of the metrics server deployment
+// by updating the pod template restart annotation. This works with any number of replicas
+// and ensures zero-downtime restarts for multi-replica deployments.
+//
+// The function name is preserved for backward compatibility, but the implementation now
+// performs a rolling restart instead of deleting a single pod.
 func DeleteMetricsServerPod(ctx context.Context, metricsServerNamespace string, logger logr.Logger, cl client.Client) error {
```
Comment (linkvt) [inline internal/controller/keda/util/util.go]: We could rename this as this isn't really deleting a metrics pod anymore, what do you think? Maybe RestartMetricsServerDeployment or so?
I don't think we need to preserve anything here for backward compatibility as we own the full code base?

File: internal/controller/keda/util/util.go [hunk 8]
```diff
@@ -5,22 +5,27 @@ import (
...
 	kedav1alpha1 "github.com/kedacore/keda-olm-operator/api/keda/v1alpha1"
 )
 
 const (
-	metricsServerPodLabelKey   = "app"
-	metricsServerPodLabelValue = "keda-metrics-apiserver"
+	metricsServerPodLabelKey    = "app"
+	metricsServerPodLabelValue  = "keda-metrics-apiserver"
+	metricsServerDeploymentName = "keda-metrics-apiserver"
+	restartAnnotationKey        = "kubectl.kubernetes.io/restartedAt"
```
Comment (linkvt) [inline internal/controller/keda/util/util.go]: I would suggest we use a custom key here instead of reusing what kubectl rollout restart does, maybesomething like `"keda.sh/restartedAt"`

File: internal/controller/keda/kedacontroller_controller.go [hunk 9]
```diff
@@ -1221,6 +1288,10 @@ func (r *KedaControllerReconciler) installMetricsServer(ctx context.Context, log
 		transforms = append(transforms, transform.ReplaceMetricsServerLogLevel(instance.Spec.MetricsServer.LogLevel, r.Scheme, logger))
 	}
 
+	if instance.Spec.MetricsServer.Replicas != nil {
+		transforms = append(transforms, transform.ReplaceReplicas(instance.Spec.MetricsServer.Replicas, r.Scheme))
+	}
```
[the remark this answers is not in the document]
Comment (linkvt) [reply internal/controller/keda/kedacontroller_controller.go:1231]: Seems to be resolved, let's see if Copilot resolves this on a rerun.

**Review by linkvt — COMMENTED** [review]
Thanks for the updates, looks better! I still had a few comments but nothing major 🙂

File: internal/controller/keda/util/util.go (hunk 7 above)
Comment (linkvt) [reply internal/controller/keda/util/util.go]: Oh I just checked, it seems keda already supports a cert reload so we don't need this anymore? It seems the cert reload is supported already since years.
Here is the response from my LLM after letting it investigate this:

> Metrics-server does not require a Pod restart for certificate rotation. KEDA’s adapter embeds `custom-metrics-apiserver`’s `AdapterBase` and runs its generic API server ([adapter](https://github.com/kedacore/keda/blob/v2.20.2/cmd/adapter/main.go#L50-L52), [run path](https://github.com/kedacore/keda/blob/v2.20.2/cmd/adapter/main.go#L317-L320)). `custom-metrics-apiserver` applies Kubernetes `SecureServingOptions` ([options.go](https://github.com/kubernetes-sigs/custom-metrics-apiserver/blob/v1.35.0/pkg/cmd/options/options.go#L88-L99)), which uses dynamic serving-certificate and client-CA providers ([serving.go](https://github.com/kubernetes/kubernetes/blob/v1.35.5/staging/src/k8s.io/apiserver/pkg/server/options/serving.go#L302-L310), [authentication.go](https://github.com/kubernetes/kubernetes/blob/v1.35.5/staging/src/k8s.io/apiserver/pkg/server/options/authentication.go#L197-L202)). These providers reload mounted certificate files without restarting the Pod. Kubernetes added dynamic serving-certificate reload in 2019 ([commit](https://github.com/kubernetes/kubernetes/commit/3f5fbfbfac281f40c11de2f57d58cc332affc37b)). The rolling restart therefore appears unnecessary for metrics-server certificate rotation.

File: api/keda/v1alpha1/kedacontroller_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (linkvt) [reply api/keda/v1alpha1/kedacontroller_types.go]: Ignore, nothing is advertised.

File: internal/controller/keda/util/util.go [hunk 10]
```diff
@@ -39,29 +44,46 @@ func CalculateSecretedDataCheckSum(m map[string][]byte) string {
...
-		client.MatchingLabels(selector),
-	}
-	err := cl.List(ctx, podList, opts...)
+	// Get the metrics server Deployment
+	deployment := &appsv1.Deployment{}
+	err := cl.Get(ctx, types.NamespacedName{
+		Name:      metricsServerDeploymentName,
+		Namespace: metricsServerNamespace,
+	}, deployment)
 	if err != nil {
```
[the remark this answers is not in the document]
Comment (linkvt) [reply internal/controller/keda/util/util.go]: Useless comment.

File: api/keda/v1alpha1/kedacontroller_types.go (hunk 4 above)
Comment (alaxkar) [reply api/keda/v1alpha1/kedacontroller_types.go]: Done

File: internal/controller/keda/util/util.go (hunk 7 above)
Comment (alaxkar) [reply internal/controller/keda/util/util.go]: @linkvt  Quick question: You mentioned metrics-server has dynamic cert reload. Should we remove the cert rotation restart code entirely, or keep it as a safety mechanism?

Comment (linkvt) [reply internal/controller/keda/util/util.go]: I would rather remove it, deleted code is the best code IMO :D 

We enabled coderabbit in the KEDA org, sorry to reviewbot spam you but I want to try if it works here as well.

@coderabbitai understand this comment thread, what is your state on this?  Also review this PR.

Comment (linkvt) [reply internal/controller/keda/util/util.go]: No I want you to review this PR and state your opinion on this.

Comment (linkvt) [reply internal/controller/keda/util/util.go]: @coderabbitai review this PR.

**Comment by linkvt** [comment]
@coderabbitai review

File: internal/controller/keda/util/util.go (hunk 7 above)
Comment (linkvt) [reply internal/controller/keda/util/util.go]: @coderabbitai what about this?

File: internal/controller/keda/kedacontroller_controller.go [hunk 11]
```diff
@@ -680,6 +680,10 @@ func (r *KedaControllerReconciler) installController(ctx context.Context, logger
 		transforms = append(transforms, transform.ReplaceKedaOperatorLogTimeEncoding(instance.Spec.Operator.LogTimeEncoding, r.Scheme, logger))
 	}
 
+	if instance.Spec.Operator.Replicas != nil {
+		transforms = append(transforms, transform.ReplaceReplicas(instance.Spec.Operator.Replicas, r.Scheme))
```
[the remark this answers is not in the document]
Comment (linkvt) [reply internal/controller/keda/kedacontroller_controller.go:684]: Isn't this resolved in https://github.com/kedacore/keda-olm-operator/pull/342 ?

**Review by linkvt — APPROVED** [review]
Thank you, looks good to me! I would like to wait for another review by Mike as well though if it is OK for you.

I hope the review bot activity wasn't to annoying, sorry again for that.

**Comment by alaxkar** [comment]
> Thank you, looks good to me! I would like to wait for another review by Mike as well though if it is OK for you.
> 
> I hope the review bot activity wasn't to annoying, sorry again for that.

Works for me. It was good learning for me as well, I made really simple things complicated here. Thank you so much for your time and patience here.

**Review by Fedosin — COMMENTED** [review]
Thanks for the updates, this is looking good now. The shared field on `GenericDeploymentSpec`, the CEL validation and the HTTP Add-on cleanup all came together well, and the final diff ended up nice and small.

Two things I'd like to settle before we merge. Everything else we can polish ourselves afterwards, no reason to keep this waiting.

To finally answer your question about `operator.replicas: 0`: I'd require `>= 1` there too. Scaling the operator to zero isn't the harmless maintenance switch it looks like. That deployment owns cert rotation and serves the gRPC endpoint the metrics adapter talks to, so zero replicas quietly takes the external metrics API down cluster-wide. It also lines the operator up with `metricsServer` and `admissionWebhooks`, and since this goes into the CRD it's much easier to relax the bound later than to tighten it.

The other one is @coderabbitai's [still-open comment](https://github.com/kedacore/keda-olm-operator/pull/340#discussion_r4018266912) about the installer manifests. Set `replicas`, then remove the field again, and the old value sticks around instead of reverting to the default. It isn't something you introduced, since every conditional transform in that file behaves the same way, but `replicas` is the first of those fields people will actually toggle, so it'll start showing up.

Everything else I had (a test or two, the README, the now-unused checksum controllers) is follow-up material and we can take care of it.

Thanks for your patience through all the review rounds and the bot spam.

**Comment by linkvt** [comment]
> The other one is @coderabbitai's [still-open comment](https://github.com/kedacore/keda-olm-operator/pull/340#discussion_r4018266912) about the installer manifests. Set `replicas`, then remove the field again, and the old value sticks around instead of reverting to the default. It isn't something you introduced, since every conditional transform in that file behaves the same way, but `replicas` is the first of those fields people will actually toggle, so it'll start showing up.

How is this the case if your PR https://github.com/kedacore/keda-olm-operator/pull/342 is already merged?
Seems like an AI tool did the same mistake coderabbit made at first by only checking PRs state?

About the replicas though: It seems we should enforce >= 1, the related issue also states that: https://redhat.atlassian.net/browse/AUTOSCALE-632 . 
> Verify that setting replicas to a value < 1 is rejected by validation

Personally I don't really care as nobody sets this by accident but if we want to follow the ACs from the ticket 0 should be rejected.

**Comment by alaxkar** [comment]
> Two things I'd like to settle before we merge. Everything else we can polish ourselves afterwards, no reason to keep this waiting.
> 
> To finally answer your question about `operator.replicas: 0`: I'd require `>= 1` there too. Scaling the operator to zero isn't the harmless maintenance switch it looks like. That deployment owns cert rotation and serves the gRPC endpoint the metrics adapter talks to, so zero replicas quietly takes the external metrics API down cluster-wide. It also lines the operator up with `metricsServer` and `admissionWebhooks`, and since this goes into the CRD it's much easier to relax the bound later than to tighten it.

Thanks, I don't have any strong opinion about it. I made it intentionally because sometimes I used to set up the operator replicas to 0 for troubleshooting. I can hardcode it to require `>= 1` replicas of the operator. I will do it in the next commit right away.

> The other one is @coderabbitai's [still-open comment](https://github.com/kedacore/keda-olm-operator/pull/340#discussion_r4018266912) about the installer manifests. Set `replicas`, then remove the field again, and the old value sticks around instead of reverting to the default. It isn't something you introduced, since every conditional transform in that file behaves the same way, but `replicas` is the first of those fields people will actually toggle, so it'll start showing up.

This was already fixed in PR `#342` as caught by Vincent, this was a false alarm raised by CodeRabbit.

**Review by Fedosin — APPROVED** [review]
Thanks! I rechecked this against current main.
